pkg/corerp/frontend/controller/applications/graph_util.go

```diff
@@ -497,21 +491,50 @@ func connectionsFromAPIData(resource generated.GenericResource) []*corerpv202310
 		data := corerpv20231001preview.ConnectionProperties{}
 		err := toStronglyTypedData(connection, &data)
 		if err == nil {
+			sourceID, ok := parseSource(to.String(resource.ID), to.String(data.Source))
+			if !ok {
+				continue
+			}
+
 			entries = append(entries, &corerpv20231001preview.ApplicationGraphConnection{
-				ID:        data.Source,
-				Direction: &dir,
+				ID:        to.Ptr(sourceID),
+				Direction: to.Ptr(dir),
 			})
 		}
 	}
 
 	// Produce a stable output
 	sort.Slice(entries, func(i, j int) bool {
-		return *entries[i].ID < *entries[j].ID
+		return to.String(entries[i].ID) < to.String(entries[j].ID)
 	})
 
 	return entries
 }
 
+func parseSource(parentID, source string) (string, bool) {
+	id, err := resources.Parse(source)
+	if err == nil && id.IsResource() {
+		return id.String(), true
+	}
+
+	sourceURL, err := url.Parse(source)
+	if err != nil {
+		return "", false
+	}
+
+	if sourceURL.Hostname() == "" || strings.Contains(sourceURL.Hostname(), ".") {
+		return "", false
+	}
+
+	parentIDParsed, err := resources.Parse(parentID)
+	if err != nil {
+		return "", false
+	}
+
+	sourceID := strings.Join([]string{parentIDParsed.RootScope(), resources.ProvidersSegment, parentIDParsed.Type(), sourceURL.Hostname()}, resources.SegmentSeparator)
+	return sourceID, true
```

Comment:
Does this assume that the resource being connected to is in the same scope? So if I use multiple resource groups this will fall a part?

Reply:
Also, what if I have a connection to something outside of Radius? eg: `http://example.com`.

Reply:
I thought that we would support only `<scheme>://<resourcename>` format for connection, so the parser is designed for the case. To support the external endpoint, I would modify the connection format like below:

```json
"connections": [
    {
        "direction": "Inbound | Outbound",
        // "id": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/containers/backendapp"
         "target": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/containers/backendapp"
         "type": "Resource | URL"
    }
],
```

instead of having `id`, we can have `target` and `type` (or something else) to define connection. Putting `url` on `id` field doesn't seem right.

cc/ @nithyatsu @rynowak

Reply:
Design doc for the features that connections support are here: https://github.com/radius-project/radius/issues/6375 @nithyatsu is the best resource for what parts of this have been implemented so far. 

I don't think we discussed yet how an "external" URL should be represented in the app graph API, so I'm open to suggestions. 

I want to make sure that we can represent a connection from `container a -> container b` and translate from URL -> resource ID in the app graph. The design for service-discovery was chosen to a support this.

Reply:
I found the bug. The following direction must be `OutBound`.

https://github.com/radius-project/radius/blob/fc4594ecdf5f2a9fc818739e925f10506c6cf980/pkg/corerp/frontend/controller/applications/graph_util.go#L496

@nithyatsu @rynowak Please review the new testdata to see if each resource connection has the right direction.